Make kubelet use an arch-specific pause image depending on GOARCH

luxas opened the pull request:
Related to: #22876, #22683 and #15140
@ixdy @pwittrock @brendandburns @mikedanese @yujuhong @thockin @zmerlynn

mikedanese commented on pkg/kubelet/types/constants.go:
```diff
@@ -16,8 +16,22 @@ limitations under the License.
+	if runtime.GOARCH == "amd64" {
+		return DefaultPodInfraContainerImageName + ":" + DefaultPodInfraContainerImageVersion
+	} else {
+		return DefaultPodInfraContainerImageName +  "-" + runtime.GOARCH + ":" + DefaultPodInfraContainerImageVersion
```
I know this pattern was not introduced here but I'm confused as to why we have been putting arch in the image name and not the build tag. It seems like it should be:
```
gcr.io/google_containers/hyperkube:v1.2.0
gcr.io/google_containers/hyperkube:v1.2.0-386
gcr.io/google_containers/hyperkube:v1.2.0-amd64
gcr.io/google_containers/hyperkube:v1.2.0-arm
```

luxas replied:
Ask @brendandburns :) 
I'm just following the conventions, but I do prefer `binary-arch:version` too, and it's implemented at quite many places atm.
I think we should go with that, but that's just my opinion :)

mikedanese replied:
I'll try to convince @david-mcmahon next time i see him :)

luxas replied:
Haha, make sure to have an issue or something if you're planning to change that, so you're listening to the community.

However, the important thing is consistency. We must choose either and go with that all the time. `binary-arch:version` is already implemented in the codebase and for example arm images are pushed with every release with `binary-arch:version` naming, so I don't know if it's worth the effort to change.

mikedanese replied:
I think our consistent should be consistent with other projects. Multi architecture support is recent enough IMO that we could still change it.

luxas replied:
Yep, it's recent enough to change, so if you think it's worth the effort and makes k8s easier to use, let's talk. 
But I would like that the `-amd64` prefix is there to allow for easy `-${ARCH}` variable passing. 
I don't think it should be `gcr.io/google_containers/hyperkube:v1.2.0` for amd64.

vishh replied:
nit: Even with the existing pattern, I'd prefer tagging `amd64` images explicitly.

luxas replied:
Yeah, it's work in progress. Images like `etcd`, `hyperkube`, `kube-proxy` are turned into `etcd-amd64`, `hyperkube-amd64` and `kube-proxy-amd64`. The older naming convention (`hyperkube` is kept for backwards-compability for a while though)

david-mcmahon replied:
@mikedanese What do you have to convince me of?   I wasn't involved in any of this naming.  My general guidance here is to follow a (sane) precedent and to be consistent with other projects.

luxas replied:
So do you vote for `binary-arch:version` or `binary:version-arch`? 
And no, you weren't involved in the naming, but now you are :)

david-mcmahon replied:
What is typical for docker containers / other projects?

luxas replied:
Docker at least uses `armhf/`, `armel/`, `ppc64le/` and so on for the "official" images. 
Not sure if that's related to this, but anyway

luxas commented:
As #15140 points out, there is many places where the pause image should be standardized and there should be a constant somewhere that the whole codebase should use, but let's take one problem at a time.

vishh commented on pkg/kubelet/types/constants.go:
```diff
@@ -16,8 +16,22 @@ limitations under the License.
 	// system default DNS resolver configuration
 	ResolvConfDefault = "/etc/resolv.conf"
+
+	DefaultPodInfraContainerImageName = "gcr.io/google_containers/pause"
```
nit: I'd prefer this logic staying out of the kubelet package. `cmd/kubelet` seems like a better location

mikedanese replied:
I agree. we should toggle this when we create the flag so it shows up in --help.

https://github.com/kubernetes/kubernetes/blob/master/cmd/kubelet/app/options/options.go#L154

luxas replied:
So where should I put it? `cmd/kubelet/app/constants.go`, `cmd/kubelet/app/pause.go` or `cmd/kubelet/app/options/options.go`, which do you prefer?

Another option would be breaking out all pause constants to some other place not directly tied to `kubelet`.

luxas commented:
@vishh Updated. PTAL

vishh commented on pkg/kubelet/types/constants.go:
```diff
@@ -17,7 +17,7 @@ limitations under the License.
 
 const (
-	PodInfraContainerImage = "gcr.io/google_containers/pause:2.0"
+
```
nit: Remove this empty line.

vishh commented on cmd/kubelet/app/options/options.go:
```diff
@@ -36,8 +37,20 @@ import (
 	experimentalFlannelOverlay = false
+
+	defaultPodInfraContainerImageName    = "gcr.io/google_containers/pause"
+	defaultPodInfraContainerImageVersion = "2.0"
```
Should we give a flag to override this image config? Other people have wanted the ability to specify custom images. Its ok to not add the flag in this PR...

luxas replied:
As I understand it, it's still possible to change the value by passing `--pod-infra-container-image` to `kubelet`. This is just the default.

Or am I missing something?

vishh replied:
My bad. Misconstrued the logic. lgtm

luxas commented:
@vishh Fixed the nit. Does this look good to you now?

vishh commented:
LGTM.

vishh commented:
@mikedanese: Do you have any other comments on this PR? If not, this is ready to be merged.

mikedanese commented:
Even though I despise the image naming convention we've settled on (probably decided by brendan and joe years ago), I will most likely survive. lgtm

mikedanese commented:
needs an update-generated-docs to update man pages

luxas commented:
`hack/update-generated-docs.sh` didn't update anything and `hack/verify-generated-docs.sh` returned `Generated docs are up to date.`.

Should I run something else?

mikedanese commented:
Oops bad guess:
```
FAIL	k8s.io/kubernetes/pkg/kubelet/dockertools [build failed]
```

luxas committed:
Make kubelet use an arch-specific pause image depending on GOARCH

luxas commented:
This works nicely on an ARM host, I tested it now.
```
$ kubelet --help
...
--pod-infra-container-image="gcr.io/google_containers/pause-arm:2.0": The image whose network/ipc namespaces containers in each pod will use.
...
```
And it starts `pause-arm` containers automatically.
On `amd64` hosts, the default is still `gcr.io/google_containers/pause:2.0`

thockin commented:
Before we commit this:

Mike,

Why?  Can you explain any value or cost to one over the other?  To my eyes
foo:1.2-arm and foo:1.3-amd64 are two point releases in the same series (as
per semver-ish semantics).  That's wrong.  OTOH foo-arm:1.2 and
foo-amd64:1.2 are clearly apples and oranges.  I still dislike the ad-hoc
nature of it.

Are there any other projects with multi-arch outputs?  What are they
doing?  What is Docker itself doing?
On Mar 25, 2016 11:11 AM, "Mike Danese" <notifications@github.com> wrote:

> Even though I despise the image naming convention we've settled on
> (probably decided by brendan and joe years ago), I will most likely
> survive. lgtm
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/23059#issuecomment-201398975>
>

luxas commented:
> Are there any other projects with multi-arch outputs?

I'm sorry but I don't think there is any other large open-source Go project that is multiarch and compareable to Kubernetes.
[Sample Github search](https://github.com/search?l=&o=desc&q=language%3AGo&ref=advsearch&s=stars&type=Repositories&utf8=%E2%9C%93)

> What is Docker itself doing?

https://github.com/docker/docker/issues/8256

thockin commented:
From that docker issue, do I understand that it is largely undecided yet?

I personally think arch has higher precedence than version, so given only
two fields (name and version) this belongs in the name.
On Mar 26, 2016 2:14 PM, "Lucas Käldström" <notifications@github.com> wrote:

> Are there any other projects with multi-arch outputs?
>
> I'm sorry but I don't think there is any other large open-source Go
> project that is multiarch and compareable to Kubernetes.
> Sample Github search
> <https://github.com/search?l=&o=desc&q=language%3AGo&ref=advsearch&s=stars&type=Repositories&utf8=%E2%9C%93>
>
> What is Docker itself doing?
>
> docker/docker#8256 <https://github.com/docker/docker/issues/8256>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/23059#issuecomment-201930309>
>

luxas commented:
Yes, I also see it as largely undecided, and I think it will take a while before docker really implement something usable.

I think these images should be multiarch primarily:
Get docker-multinode working:
 - gcr.io/google_containers/hyperkube, pending #21617
 - gcr.io/google_containers/etcd, fixed in #19216
 - gcr.io/google_containers/flannel, pending #21617
 - gcr.io/google_containers/pause, pending #23009

Get DNS working:
 - gcr.io/google_containers/kube2sky, pending #19376
 - gcr.io/google_containers/skydns, pending #19376
 - gcr.io/google_containers/exechealthz

Docker wrapped Kubernetes binaries:
 - gcr.io/google_containers/kube-proxy, fixed in #19769
 - gcr.io/google_containers/kube-apiserver, fixed in #19769
 - gcr.io/google_containers/kube-controller-manager, fixed in #19769
 - gcr.io/google_containers/kube-scheduler, fixed in #19769
 - gcr.io/google_containers/kubelet, doesn't really exist yet: #19177

I worked with the dashboard team, and it has had multiarch support from the first release.
I'm maybe gonna add multiarch support for official heapster too, now when it has reached v1.
On [kubernetes-on-arm](http://bit.ly/1U5nBgx), I already have heapster, influxdb and grafana nicely running as a cluster addon.

When we once has ported Kubernetes to ARM and decided about the naming, it's much easier to add support for ARM 64-bit, ppc64le and maybe some other arch, who knows?

luxas commented:
@thockin @mikedanese @vishh Can we proceed with this one?
This PR isn't that tightly integrated with the naming as it uses two already existing images as the default (`pause` and `pause-arm`)
It this ok?

vishh commented:
LGTM

vishh commented:
@thockin 

> I personally think arch has higher precedence than version, so given only
two fields (name and version) this belongs in the name.

+1
